pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/UGICache.java

```diff
@@ -0,0 +1,129 @@
+    private static Map<SegmentTransactionId, TimedProxyUGI> cache = new ConcurrentHashMap<>();
+    //private static DelayQueue<TimedProxyUGI> delayQueue = new DelayQueue<>();
+    private static DelayQueue<TimedProxyUGI>[] delayQueues = new DelayQueue<>[64];
+    public static long UGI_CACHE_EXPIRY = 15 * 1 * 1000L; // 15 Minutes
+
+    public UGICache() {
+        for (int i = 0; i < delayQueues.length; i++) {
+            delayQueues[i] = new DelayQueue<>();
+        }
+    }
+
+    public TimedProxyUGI getTimedProxyUGI(String user, SegmentTransactionId session) throws IOException {
+
+        Integer segmentId = session.getSegmentId();
+        synchronized (delayQueues[segmentId]) {
+            // use the opportunity to cleanup any expired entries
+            cleanup(segmentId);
+            TimedProxyUGI timedProxyUGI = cache.get(session);
+            if (timedProxyUGI == null) {
+                LOG.info(session.toString() + " Creating proxy user = " + user);
+                UserGroupInformation proxyUGI =
+                        UserGroupInformation.createProxyUser(user, UserGroupInformation.getLoginUser());
+                timedProxyUGI = new TimedProxyUGI(proxyUGI, session);
+                delayQueues[segmentId].offer(timedProxyUGI);
+                cache.put(session, timedProxyUGI);
+            }
+            timedProxyUGI.incrementCounter();
+            return timedProxyUGI;
+        }
+    }
+
+    private cleanup(Integer segmentId) {
+        // poll segment expiration queue for all expired entries and clean them if possible   
+        TimedProxyUGI ugi = null;
+        while ((ugi = delayQueues[segmentId].poll()) != null) {
+            // place it back in the queue if still in use and was not closed
+            if (!closeUGI(ugi)) {
+                delayQueues[segmentId].offer(ugi);
+            }
+            LOG.info("Delay Queue Size for segment " +
+                    segmentId + " = " + delayQueues[segmentId].size());
+        }
+    }
+    
+    public void release(TimedProxyUGI timedProxyUGI, boolean forceClean) {
+
+        Integer segmentId = timedProxyUGI.getSession().getSegmentId();
+        
+        timedProxyUGI.resetTime();
+        timedProxyUGI.decrementCounter();
+        
+        if (forceClean) {
+            synchronized (delayQueues[segmentId]) {
+                timedProxyUGI.setExpired();
+                closeUGI(timedProxyUGI);
+            }
+        }
+    }
+
+    private static boolean closeUGI(TimedProxyUGI expiredProxyUGI) {
+
+        SegmentTransactionId session = expiredProxyUGI.getSession();
+        Integer segmentId = session.getSegmentId();
+        //synchronized (delayQueues[segmentId]) {
+            String fsMsg = "FileSystem for proxy user = " + expiredProxyUGI.getProxyUGI().getUserName();
+            try {
+                // The UGI object is still being used by another thread
+                if (expiredProxyUGI.inProgress.get() != 0) {
```

Comment:
Should this be `expiredProxyUGI.getCounter()`? Maybe there should be a method `isInUse()` that does the comparison to `0` too...